File: pkg/adf/sourcemap.go
// SPDX-FileCopyrightText: (c) 2026 Rafal Zajac
// SPDX-License-Identifier: MIT

package adf

import (
	"fmt"
	"strings"
)

// Span is the half-open byte range [Start, End) that a rendered block occupies
// in the Markdown produced by [ADF.MarshallMarkdownMapped]. The offsets index
// the returned byte slice, frontmatter included.
type Span struct {
	Start int
	End   int
}

// Origin links one rendered top-level block back to the ADF node that produced
// it. It is the invisible anchor that makes push possible: on push, cfsync
// re-renders the cached ADF, aligns the user's edited Markdown blocks against
// those origins by normalized content (LCS via [diffBlocks]), and rebuilds the
// ADF from the cached tree plus the expressed edits using [Origin.NodeIndex].
// No marker is written into the Markdown itself.
type Origin struct {
	// NodeIndex is the position, in the document's top-level Content slice, of the
	// source node. Blocks that render to nothing have no origin, so indices are
	// not necessarily contiguous. Put matches edits by content and applies them
	// through this index.
	NodeIndex int

	// Type is the source node's ADF type, such as "paragraph" or "table".
	Type string

	// LocalID is the source node's localId attribute, or "" when it has none.
	// It is recorded for tooling and diagnostics; push matching uses content
	// alignment, not this field.
	LocalID string

	// Span is the block's byte range in the rendered Markdown.
	Span Span
}

// SourceMap is the ordered origin table for one render: the byte where the body
// begins and, per non-empty top-level block, the [Origin] linking it to its
// source node. Its zero value describes a document with no rendered body.
type SourceMap struct {
	// BodyStart is the byte offset at which the rendered body begins, just after
	// the frontmatter and its separating blank line. It equals the length of the
	// output when the document has no rendered body.
	BodyStart int

	// Origins holds one entry per non-empty top-level block, in document order.
	Origins []Origin
}

// MarshallMarkdownMapped renders the document exactly as [ADF.MarshallMarkdown]
// does and additionally returns a [SourceMap] describing where each top-level
// block landed in the output and which ADF node produced it. The byte slice it
// returns is identical to what MarshallMarkdown returns for the same input; the
// map is the side table push needs to back-port edits without writing any
// anchor into the Markdown. See [Origin] for the role it plays.
func (adf *ADF) MarshallMarkdownMapped(
	assets map[string]string,
) ([]byte, *SourceMap, error) {

	return adf.marshallMapped(assets, nil)
}

// marshallMapped is [ADF.MarshallMarkdownMapped] with an optional [Links] to
// rewrite page links; a nil links renders every link unchanged.
func (adf *ADF) marshallMapped(
	assets map[string]string,
	links Links,
) ([]byte, *SourceMap, error) {

	if adf.Doc.Type != "doc" {
		return nil, nil, fmt.Errorf("root node is %q, want doc", adf.Doc.Type)
	}

	ctx := mdCtx{assets: assets, ambig: adf.ambiguousMentions(), links: links}
	blocks := renderBlockList(adf.Doc.Content, ctx)

	var b strings.Builder
	b.WriteString(adf.frontmatter(assets))

	sm := &SourceMap{}
	if len(blocks) > 0 {
		b.WriteString("\n\n")
		sm.BodyStart = b.Len()
		sm.Origins = make([]Origin, 0, len(blocks))
		for i, blk := range blocks {
			if i > 0 {
				b.WriteString("\n\n")
			}
			start := b.Len()
			b.WriteString(blk.Text)
			node := adf.Doc.Content[blk.NodeIndex]
			sm.Origins = append(sm.Origins, Origin{
				NodeIndex: blk.NodeIndex,
				Type:      node.Type,
				LocalID:   node.attrStr("localId"),
				Span:      Span{Start: start, End: b.Len()},
			})
		}
	} else {
		sm.BodyStart = b.Len()
	}
	b.WriteByte('\n')

	return []byte(b.String()), sm, nil
}

File: packages/core/src/adf/lens/sourcemap.ts
// SPDX-FileCopyrightText: (c) 2026 Rafal Zajac
// SPDX-License-Identifier: MIT

// The source map, ported from `pkg/adf/sourcemap.go`, plus the top-level render
// assembly it shares with `MarshallMarkdown`. A render pairs each non-empty
// top-level block with the ADF node that produced it, recording the block's byte
// range in the output — the invisible anchor push uses to back-port edits
// without writing any marker into the Markdown. Offsets are string indices
// (UTF-16 code units), used consistently for both the recorded spans and the
// slices that read them back, so the map stays self-consistent.
//
// Landed ahead of its plan slot (M5.2) because the Put lens (M4) needs
// `baselineBlocks`, which the source map produces.

import { type ADF, attrStr } from "../../models/adf.ts";
import type { Links } from "../links.ts";
import { type MdBlock, newBlock } from "../parse/blocks.ts";
import {
    ambiguousMentions,
    frontmatter,
    goQuote,
} from "../render/frontmatter.ts";
import { type MdCtx, renderBlockList } from "../render/markdown.ts";

/**
 * Span is the half-open range `[start, end)` a rendered block occupies in the
 * Markdown produced by {@link marshallMapped}. The offsets index the returned
 * string, frontmatter included.
 */
export interface Span {
    start: number;
    end: number;
}

/**
 * Origin links one rendered top-level block back to the ADF node that produced
 * it. On push, the cached ADF is re-rendered, the user's edited blocks are
 * aligned against these origins by normalized content, and the ADF is rebuilt
 * from the cached tree plus the expressed edits using {@link Origin.nodeIndex}.
 */
export interface Origin {
    /**
     * The position, in the document's top-level content slice, of the source
     * node. Blocks that render to nothing have no origin, so indices are not
     * necessarily contiguous.
     */
    nodeIndex: number;
    /** The source node's ADF type, such as `paragraph` or `table`. */
    type: string;
    /** The source node's `localId`, or `""` when it has none. */
    localId: string;
    /** The block's range in the rendered Markdown. */
    span: Span;
}

/**
 * SourceMap is the ordered origin table for one render: the offset where the
 * body begins and, per non-empty top-level block, the {@link Origin} linking it
 * to its source node.
 */
export interface SourceMap {
    /**
     * The offset at which the rendered body begins, just after the frontmatter
     * and its separating blank line. When the document has no rendered body it
     * points just past the frontmatter's closing `---`, one byte short of the
     * output length — the render still appends a single trailing newline after it.
     */
    bodyStart: number;
    /** One entry per non-empty top-level block, in document order. */
    origins: Origin[];
}

/**
 * marshallMapped renders the document as {@link marshallMarkdownAssets} does and
 * additionally returns a {@link SourceMap} describing where each top-level block
 * landed and which ADF node produced it. It throws when the root is not a `doc`.
 * `margin` sets the soft-wrap column (0, the default, disables wrapping); it only
 * affects text written to a note, never the reflow-agnostic block diff.
 */
export function marshallMapped(
    adf: ADF,
    assets: Record<string, string>,
    links: Links | null,
    margin = 0,
): [string, SourceMap] {
    if (adf.doc.type !== "doc") {
        throw new Error(`root node is ${goQuote(adf.doc.type)}, want doc`);
    }
    const ctx: MdCtx = { assets, ambig: ambiguousMentions(adf), links, margin };
    const blocks = renderBlockList(adf.doc.content ?? [], ctx);

    let b = frontmatter(adf, assets);
    const sm: SourceMap = { bodyStart: 0, origins: [] };
    const content = adf.doc.content ?? [];
    if (blocks.length > 0) {
        b += "\n\n";
        sm.bodyStart = b.length;
        for (const [i, blk] of blocks.entries()) {
            if (i > 0) {
                b += "\n\n";
            }
            const start = b.length;
            b += blk.text;
            const node = content[blk.nodeIndex] ?? { type: "" };
            sm.origins.push({
                nodeIndex: blk.nodeIndex,
                type: node.type,
                localId: attrStr(node.attrs, "localId"),
                span: { start, end: b.length },
            });
        }
    } else {
        sm.bodyStart = b.length;
    }
    b += "\n";
    return [b, sm];
}

/**
 * marshallMarkdownMapped renders the document and returns it with its
 * {@link SourceMap}, mirroring Go's `ADF.MarshallMarkdownMapped`. The string is
 * identical to what {@link marshallMarkdownAssets} returns for the same input.
 */
export function marshallMarkdownMapped(
    adf: ADF,
    assets: Record<string, string>,
): [string, SourceMap] {
    return marshallMapped(adf, assets, null);
}

/**
 * marshallMarkdownAssets renders the document with resolved image assets and no
 * link rewriting, mirroring Go's `ADF.MarshallMarkdown`.
 */
export function marshallMarkdownAssets(
    adf: ADF,
    assets: Record<string, string>,
): string {
    return marshallMapped(adf, assets, null)[0];
}

/**
 * marshallMarkdownLinks renders as {@link marshallMarkdownAssets} does, also
 * rewriting each link to a pulled Confluence page into its local Markdown link
 * via `links`. Mirrors Go's `ADF.MarshallMarkdownLinks`. `margin` sets the
 * soft-wrap column (0, the default, disables wrapping).
 */
export function marshallMarkdownLinks(
    adf: ADF,
    assets: Record<string, string>,
    links: Links | null,
    margin = 0,
): string {
    return marshallMapped(adf, assets, links, margin)[0];
}

/**
 * baselineBlocks renders the document and returns its top-level blocks, each
 * paired with the {@link Origin} linking it to the ADF node that produced it. It
 * is the authoritative baseline for a push diff: the block text comes straight
 * from the render and every block carries its source node index and localId.
 */
export function baselineBlocks(
    adf: ADF,
    assets: Record<string, string>,
    links: Links | null,
): [MdBlock[], Origin[]] {
    const [md, sm] = marshallMapped(adf, assets, links);
    const blocks = sm.origins.map((o) =>
        newBlock(md.slice(o.span.start, o.span.end)),
    );
    return [blocks, sm.origins];
}
